Added check-rabbitmq-node-usage.rb plugin (built on top of check-rabb…

DvaBearqLoza opened the pull request:
Hi,

I have added a new plugin, check-rabbitmq-node-usage.rb, which was based on check-rabbitmq-node-health.rb.
Due to the amount of changes, I thought it would be better to push new plugin, instead of changing the old one.

There were issues for this:
https://github.com/sensu-plugins/sensu-plugins-rabbitmq/issues/44
https://github.com/sensu-plugins/sensu-plugins-rabbitmq/issues/45

What was added/changed:
The script was only checking for mem/fd/socket. I have added checks for proc and disk usage.
Also, instead of rest_client which is deprecated, I have used carrot-top (as requested in the issue #45).

Due to the  structure of the script, it was only checking until it reached the first issue. For the needs of our company, I changed it so that we have the "type" option, with which you pick one resource (mem,fd,socket,proc,disk) to be checked.
Also, if the usage is under the warning level, the return message will show usage anyway.

I have made some corrections to the comments, as well as added some new ones.
I added usage the comment with the example too.

eheydrick commented:
There are merge conflicts, can you rebase.

DvaBearqLoza commented:
Hey :)
The conflict was with the ChangeLog.
There was a new upload, therefore my pull request which was earlier created a conflict.
I have just cleaned the ChangeLog to be as it was.

What I added originally was:

## [2.0.1] - 2016-10-28
### Added
 - check-rabbitmq-node-usage.rb plugin

Could you make the addition to the ChangeLog as you see fit please?

majormoses commented:
Thank you for your contribution! Any chance we could add some unit testing for the new plugin?

DvaBearqLoza commented:
@majormoses Hi :) 
What kind of unit testing are you thinking of?

majormoses commented:
If you could mock the response you would get from RMQ and just test that with the right inputs from your mocks you get the right output with something like rspec. I will review it in a sec.

majormoses commented on bin/check-rabbitmq-node-usage.rb:
```diff
@@ -0,0 +1,306 @@
+         long: '--type TYPE',
+         in: %w(mem socket fd proc disk)
+
+  option :memwarn,
```
if you are specifying type is there some reason to not just have standard warnings and crits? Whats the value added here?

majormoses commented:
I also do not see an entry in changelog for this.

DvaBearqLoza commented:
In the first comment I have left some details as to why changes have been made. The needs for the check we are using werent met, and I made some changes, which were enough to justify creating a separate plugin.
We are currently using this check in production, and it has been tested multiple times, and is working properly.
Regarding your question about specifying a type, I would have to go again through the plugins, as I have done that sometime in Oct. 2016, and can barely remember half of it :)

As for the changelog, some uploads were made, which happened after my pull request, and created conflict in the changelog. Since long time has passed since I got a reply regarding this, I left the comment:

"Hey :)
The conflict was with the ChangeLog.
There was a new upload, therefore my pull request which was earlier created a conflict.
I have just cleaned the ChangeLog to be as it was.

What I added originally was:

[2.0.1] - 2016-10-28

Added

check-rabbitmq-node-usage.rb plugin
Could you make the addition to the ChangeLog as you see fit please?"

majormoses commented:
@DvaBearqLoza makes sense to me, I sometimes can't remember code I wrote yesterday let alone months ago. Honestly I am ok with merging without adding tests as it has been running in production for a while without issue. I would like to understand why all the extra params as it seems like that is really the whole point of having `type` from what I see. Understood given the time it has been sitting there some CHANGELOG conflicts are a given. I am ok with having me make the CHANGELOG notification if you can address the need of so many options.

I fail to see why specifying `--type and --mwarn` make sense, if we know we want to check memory just use a generic warn/crit threshold options. An example that maybe illustrates my point is that `--type disk and --mwarn` makes even less sense but that is something that is an option and would break I am sure.

majormoses commented:
OK I think I see the goal, it was to allow checking multiple types at once within the same instance in one check rather than specifying for each type as its own "instance". Am I right? If that's the case this certainly makes more sense.

DvaBearqLoza commented:
Yes, the idea was that you can check multiple things at the same time, if you wanted to do so.

DvaBearqLoza commented:
As for the changelog, I would appreciate if you add the changes as you see fit. The date of the change is not important as far as I am concerned.

And thank you for the input/replies/help :)

majormoses commented:
I have rebased your forks branch and fixed up the CHANGELOG entry.

majormoses commented on bin/check-rabbitmq-node-usage.rb:
```diff
@@ -0,0 +1,306 @@
+
+    { 'status' => output[:status], 'message' => output[:message] }
+  rescue Errno::ECONNREFUSED => e
+    { 'status' => 'critical', 'message' => e.message }
```
we already have checks to determine if a node is alive `check-rabbitmq-alive.rb`, this should be an unknown to reduce the number of parallel alerts.

DvaBearqLoza commented:
I have made the changes. Hopefully, all is good now.
This was my first pull request, so thanks again for the input and help.

majormoses commented:
@DvaBearqLoza yes it looks good now, I rebased against upstreams master to remove the duplicate changelog entries. as soon as CI passes I will merge. If I don't do release within 24 hours please open an issue and ping me.

majormoses commented:
https://rubygems.org/gems/sensu-plugins-rabbitmq/versions/2.3.0
